#### Constructor Standings

What is the Constructors’ points system?
Apart from the drivers’ championship title, the teams simultaneously also battle for the constructors’ title in a championship season.

The points system for the constructors’ titles is the same as the drivers’ titles. Although, for constructors’ points, the teams add up the points bagged by both of their drivers. The higher a team finishes, the more money they have at their disposal in the next year’s championship.

The prize pot that is distributed around the grid in 2023 based on the final year's standings is worth a total of more than $900million — of a total F1 prize pot of around $2.2bn.

Nowadays, the top team receives 14% of the total prize pot, whereas the bottom team receives 6% of it. Previously, the top team used to receive 20% and the bottom team received only 4%.

This agreement is locked in until January 2025.

| Position     | Points Scored |
| ----------- | ----------- |
| 1           |      25     |
| 2           | 18     |
| 3          | 15   |
| 4          | 12     |
| 5          | 10     |
| 6           | 8     |
| 7          | 6    |
| 8         | 4    |
| 9        | 2    |
| 10           | 1     |


In [1]:
# importing required libraries 

import pandas as pd
import seaborn as sns
import plotly.express as px
import plotly.figure_factory as ff
import plotly.graph_objects as go
import matplotlib.pyplot as plt 
import numpy as np
from sklearn.model_selection import train_test_split
import warnings
warnings.simplefilter("ignore")
pd.set_option('display.max_columns', None)

Let's see how are the Constructor Standings from 2012.

In [2]:
constructor_standings = pd.read_csv("data/constructor_standings.csv")
constructor = pd.read_csv("data/constructors.csv")
#constructor_results = pd.read_csv("data/constructor_results.csv")
races = pd.read_csv("data/races.csv")

#### Constructor Standings

In [3]:
constructor_standings.head()

constructorStandingsId  raceId  constructorId  points  position   
0                       1      18              1    14.0         1  \
1                       2      18              2     8.0         3   
2                       3      18              3     9.0         2   
3                       4      18              4     5.0         4   
4                       5      18              5     2.0         5   

  positionText  wins  
0            1     1  
1            3     0  
2            2     0  
3            4     0  
4            5     0

In [4]:
constructor_standings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12941 entries, 0 to 12940
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   constructorStandingsId  12941 non-null  int64  
 1   raceId                  12941 non-null  int64  
 2   constructorId           12941 non-null  int64  
 3   points                  12941 non-null  float64
 4   position                12941 non-null  int64  
 5   positionText            12941 non-null  object 
 6   wins                    12941 non-null  int64  
dtypes: float64(1), int64(5), object(1)
memory usage: 707.8+ KB


We have 7 columns, 12941 rows and no Nan values.

Now, let's see the constructor dataframe.

Let's now drop the columns we don't need:

In [5]:
# List of columns to drop
columns_to_drop = ['constructorStandingsId', 'positionText']
# Drop the specified columns
constructor_standings.drop(columns_to_drop, axis=1, inplace=True)

# Print the updated DataFrame
constructor_standings.head()

raceId  constructorId  points  position  wins
0      18              1    14.0         1     1
1      18              2     8.0         3     0
2      18              3     9.0         2     0
3      18              4     5.0         4     0
4      18              5     2.0         5     0

#### Constructor dataframe

In [6]:
constructor.head()

constructorId constructorRef        name nationality   
0              1        mclaren     McLaren     British  \
1              2     bmw_sauber  BMW Sauber      German   
2              3       williams    Williams     British   
3              4        renault     Renault      French   
4              5     toro_rosso  Toro Rosso     Italian   

                                                 url  
0               http://en.wikipedia.org/wiki/McLaren  
1            http://en.wikipedia.org/wiki/BMW_Sauber  
2  http://en.wikipedia.org/wiki/Williams_Grand_Pr...  
3  http://en.wikipedia.org/wiki/Renault_in_Formul...  
4   http://en.wikipedia.org/wiki/Scuderia_Toro_Rosso

In [7]:
constructor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211 entries, 0 to 210
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   constructorId   211 non-null    int64 
 1   constructorRef  211 non-null    object
 2   name            211 non-null    object
 3   nationality     211 non-null    object
 4   url             211 non-null    object
dtypes: int64(1), object(4)
memory usage: 8.4+ KB


In [8]:
# List of columns to drop
columns_to_drop = ['url','nationality','constructorRef']
# Drop the specified columns
constructor.drop(columns_to_drop, axis=1, inplace=True)

# Print the updated DataFrame
constructor.head()

constructorId        name
0              1     McLaren
1              2  BMW Sauber
2              3    Williams
3              4     Renault
4              5  Toro Rosso

We can merge both based on the __constructorId__ column

Since our __constructor_standings__ dataframe does not have the constructor names, let's merge it with  the __contructor__ dataframe.

In [9]:
merged_constructor_df = pd.merge(constructor_standings,constructor, on="constructorId",how="inner")

In [10]:
merged_constructor_df.head()

raceId  constructorId  points  position  wins     name
0      18              1    14.0         1     1  McLaren
1      19              1    24.0         1     1  McLaren
2      20              1    28.0         3     1  McLaren
3      21              1    34.0         3     1  McLaren
4      22              1    42.0         3     1  McLaren

In [11]:
merged_constructor_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12941 entries, 0 to 12940
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   raceId         12941 non-null  int64  
 1   constructorId  12941 non-null  int64  
 2   points         12941 non-null  float64
 3   position       12941 non-null  int64  
 4   wins           12941 non-null  int64  
 5   name           12941 non-null  object 
dtypes: float64(1), int64(4), object(1)
memory usage: 606.7+ KB


Since we still need information from the races csv file, we will rename the __name__ column to __constructor_name__ column, to avoid conflicts in the naming of the columns.

In [12]:
merged_constructor_df.rename(columns={'name': 'constructor_name'}, inplace=True)

merged_constructor_df.head()


raceId  constructorId  points  position  wins constructor_name
0      18              1    14.0         1     1          McLaren
1      19              1    24.0         1     1          McLaren
2      20              1    28.0         3     1          McLaren
3      21              1    34.0         3     1          McLaren
4      22              1    42.0         3     1          McLaren

#### Races

In [13]:
races.head()

raceId  year  round  circuitId                   name        date   
0       1  2009      1          1  Australian Grand Prix  2009-03-29  \
1       2  2009      2          2   Malaysian Grand Prix  2009-04-05   
2       3  2009      3         17     Chinese Grand Prix  2009-04-19   
3       4  2009      4          3     Bahrain Grand Prix  2009-04-26   
4       5  2009      5          4     Spanish Grand Prix  2009-05-10   

       time                                                url fp1_date   
0  06:00:00  http://en.wikipedia.org/wiki/2009_Australian_G...       \N  \
1  09:00:00  http://en.wikipedia.org/wiki/2009_Malaysian_Gr...       \N   
2  07:00:00  http://en.wikipedia.org/wiki/2009_Chinese_Gran...       \N   
3  12:00:00  http://en.wikipedia.org/wiki/2009_Bahrain_Gran...       \N   
4  12:00:00  http://en.wikipedia.org/wiki/2009_Spanish_Gran...       \N   

  fp1_time fp2_date fp2_time fp3_date fp3_time quali_date quali_time   
0       \N       \N       \N       \N       \N         \N         \N  \
1       \N       \N       \N       \N       \N         \N         \N   
2       \N       \N       \N       \N       \N         \N         \N   
3       \N       \N       \N       \N       \N         \N         \N   
4       \N       \N       \N       \N       \N         \N         \N   

  sprint_date sprint_time  
0          \N          \N  
1          \N          \N  
2          \N          \N  
3          \N          \N  
4          \N          \N

In [14]:
races.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1102 entries, 0 to 1101
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   raceId       1102 non-null   int64 
 1   year         1102 non-null   int64 
 2   round        1102 non-null   int64 
 3   circuitId    1102 non-null   int64 
 4   name         1102 non-null   object
 5   date         1102 non-null   object
 6   time         1102 non-null   object
 7   url          1102 non-null   object
 8   fp1_date     1102 non-null   object
 9   fp1_time     1102 non-null   object
 10  fp2_date     1102 non-null   object
 11  fp2_time     1102 non-null   object
 12  fp3_date     1102 non-null   object
 13  fp3_time     1102 non-null   object
 14  quali_date   1102 non-null   object
 15  quali_time   1102 non-null   object
 16  sprint_date  1102 non-null   object
 17  sprint_time  1102 non-null   object
dtypes: int64(4), object(14)
memory usage: 155.1+ KB


Drop all columns we don't need:

In [15]:
# List of columns to drop
columns_to_drop = ['round','date','time','url','fp1_date','fp1_time','fp2_date','fp2_time','fp3_date','fp3_time','quali_date','quali_time','sprint_date','sprint_time']
# Drop the specified columns
races.drop(columns_to_drop, axis=1, inplace=True)

races.head()

raceId  year  circuitId                   name
0       1  2009          1  Australian Grand Prix
1       2  2009          2   Malaysian Grand Prix
2       3  2009         17     Chinese Grand Prix
3       4  2009          3     Bahrain Grand Prix
4       5  2009          4     Spanish Grand Prix

In [16]:
races.rename(columns={'name': 'circuit_name'}, inplace=True)

races.head()

raceId  year  circuitId           circuit_name
0       1  2009          1  Australian Grand Prix
1       2  2009          2   Malaysian Grand Prix
2       3  2009         17     Chinese Grand Prix
3       4  2009          3     Bahrain Grand Prix
4       5  2009          4     Spanish Grand Prix

We can now merge the __races__ dataframe:

In [17]:
merged_constructor_races_df = pd.merge(merged_constructor_df,races, on="raceId",how="inner")

In [18]:
merged_constructor_races_df.head()

raceId  constructorId  points  position  wins constructor_name  year   
0      18              1    14.0         1     1          McLaren  2008  \
1      18              2     8.0         3     0       BMW Sauber  2008   
2      18              3     9.0         2     0         Williams  2008   
3      18              4     5.0         4     0          Renault  2008   
4      18              5     2.0         5     0       Toro Rosso  2008   

   circuitId           circuit_name  
0          1  Australian Grand Prix  
1          1  Australian Grand Prix  
2          1  Australian Grand Prix  
3          1  Australian Grand Prix  
4          1  Australian Grand Prix

In [19]:
merged_constructor_races_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12941 entries, 0 to 12940
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   raceId            12941 non-null  int64  
 1   constructorId     12941 non-null  int64  
 2   points            12941 non-null  float64
 3   position          12941 non-null  int64  
 4   wins              12941 non-null  int64  
 5   constructor_name  12941 non-null  object 
 6   year              12941 non-null  int64  
 7   circuitId         12941 non-null  int64  
 8   circuit_name      12941 non-null  object 
dtypes: float64(1), int64(6), object(2)
memory usage: 910.0+ KB


In [20]:
merged_constructor_races_df.year.unique()

array([2008, 2007, 2006, 2005, 2004, 2003, 2002, 2001, 2000, 1999, 1998,
       1997, 1996, 1995, 1994, 1993, 1992, 1990, 2009, 1991, 1989, 1988,
       1987, 1986, 1985, 1984, 1983, 1982, 1981, 1980, 1979, 1978, 1977,
       1976, 1975, 1974, 1973, 1972, 1971, 1968, 2010, 2011, 2012, 2013,
       2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 1967,
       1966, 1965, 1963, 1964, 1962, 1961, 1960, 1959, 1958, 1970, 1969])

In [21]:
merged_constructor_races_df.constructor_name.unique()

array(['McLaren', 'BMW Sauber', 'Williams', 'Renault', 'Toro Rosso',
       'Ferrari', 'Toyota', 'Red Bull', 'Honda', 'Force India',
       'Super Aguri', 'Spyker', 'MF1', 'Spyker MF1', 'Jordan', 'BAR',
       'Sauber', 'Minardi', 'Jaguar', 'Arrows', 'Benetton', 'Prost',
       'Stewart', 'Tyrrell', 'Lola', 'Ligier', 'Footwork', 'Forti',
       'Pacific', 'Simtek', 'Larrousse', 'Team Lotus', 'Brabham',
       'Dallara', 'March', 'Fondmetal', 'Coloni', 'Leyton House', 'AGS',
       'Euro Brun', 'Osella', 'Onyx', 'Life', 'Brawn', 'Lambo', 'Rial',
       'Zakspeed', 'RAM', 'Alfa Romeo', 'Spirit', 'Toleman', 'ATS',
       'Theodore', 'Fittipaldi', 'Ensign', 'Shadow', 'Merzario', 'Wolf',
       'Brabham-Alfa Romeo', 'Kauhsen', 'Rebaque', 'Brabham-Ford',
       'Hesketh', 'Surtees', 'Martini', 'BRM', 'Penske', 'LEC', 'McGuire',
       'Apollon', 'Boro', 'Kojima', 'Parnelli', 'Maki', 'Shadow-Ford',
       'Embassy Hill', 'Lyncar', 'Shadow-Matra', 'Iso Marlboro', 'Amon',
       'Trojan', 'Toke

Let's now create our plots:

The distribution of the constructor standings from 2012 until 2022

In [22]:
constructor_standings_2012_to_2022 = merged_constructor_races_df[
    (merged_constructor_races_df['year'] >= 2012) & (merged_constructor_races_df['year'] <= 2022)
]

# Group the data by constructor_name and calculate the mean position for each constructor
constructor_standings_mean = constructor_standings_2012_to_2022.groupby('constructor_name')['position'].mean()

# Sort the data in ascending order based on mean position
constructor_standings_mean = constructor_standings_mean.sort_values(ascending=False)

# Calculate the percentage relative to the best possible mean position (100% for the best)
best_mean_position = constructor_standings_mean.min()
constructor_standings_percentage = 100 * (best_mean_position / constructor_standings_mean)

# Create a DataFrame with constructor names and mean positions in percentage
data = pd.DataFrame({'constructor_name': constructor_standings_mean.index, 'percentage': constructor_standings_percentage})

# Create a horizontal bar chart using Plotly Express
fig = px.bar(data, x='percentage', y='constructor_name',
             color_discrete_sequence=["#00A08B"],
             title='Constructor Standings from 2012 to 2022',
             labels={'percentage': 'Percentage of Best Mean Position'},
             height=600, width=800)

# Update x-axis title and fonts
fig.update_xaxes(title_text='Percentage Distribution ', title_font=dict(size=18))

# Update y-axis title
fig.update_yaxes(title_text='Constructor Name', title_font=dict(size=18))

# Update marker settings
fig.update_traces(marker=dict(line=dict(width=1, color='black')), textposition='inside')

# Update title font size
fig.update_layout(title_font=dict(size=24))

# Show the plot
fig.show()


Mercedes has been the best team.

Let's now have a closer look from the year 2021 and 2022, because __Mclaren__ is only racing for two years now.

In [23]:
constructor_standings_2021_to_2022 = merged_constructor_races_df[
    (merged_constructor_races_df['year'] >= 2021) & (merged_constructor_races_df['year'] <= 2022)
]

# Group the data by constructor_name and calculate the mean position for each constructor
constructor_standings_mean = constructor_standings_2021_to_2022.groupby('constructor_name')['position'].mean()

# Sort the data in ascending order based on mean position
constructor_standings_mean = constructor_standings_mean.sort_values(ascending=False)  # Reverse the order

# Calculate the percentage relative to the best possible mean position (100% for the best)
best_mean_position = constructor_standings_mean.min()
constructor_standings_percentage = 100 * (best_mean_position / constructor_standings_mean)

# Create a DataFrame with constructor names and mean positions in percentage
data = pd.DataFrame({'constructor_name': constructor_standings_mean.index, 'percentage': constructor_standings_percentage})

# Create a bar chart using Plotly Express
fig = px.bar(data, x='percentage', y='constructor_name', orientation='h',
             color_discrete_sequence=["#00A08B"],
             title='Constructor Standings for 2021 and 2022 (Percentage)',
             labels={'percentage': 'Performance Percentage'},
             height=600, width=800)

# Update x-axis title and fonts
fig.update_xaxes(title_text='Distribution Percentage',title_font=dict(size=18))

# Update y-axis title
fig.update_yaxes(title_text='Constructor Name',title_font=dict(size=18))

# Update marker settings
fig.update_traces(marker=dict(line=dict(width=1, color='black')))

# Update title font size
fig.update_layout(title_font=dict(size=24))


# Show the plot
fig.show()


As we can see __Red Bull__ has the best results and __Aston Martin__ appears in 8th position. 

Must definitely improve in these numbers to increase the budget for the next season.

In [24]:
constructor_standings_2021_to_2022 = merged_constructor_races_df[
    (merged_constructor_races_df['year'] == 2021)
]

# Group the data by constructor_name and calculate the mean position for each constructor
constructor_standings_mean = constructor_standings_2021_to_2022.groupby('constructor_name')['position'].mean()

# Sort the data in ascending order based on mean position
constructor_standings_mean = constructor_standings_mean.sort_values(ascending=False)  # Reverse the order

# Calculate the percentage relative to the best possible mean position (100% for the best)
best_mean_position = constructor_standings_mean.min()
constructor_standings_percentage = 100 * (best_mean_position / constructor_standings_mean)

# Create a DataFrame with constructor names and mean positions in percentage
data = pd.DataFrame({'constructor_name': constructor_standings_mean.index, 'percentage': constructor_standings_percentage})

# Create a bar chart using Plotly Express
fig = px.bar(data, x='percentage', y='constructor_name', orientation='h',
             color_discrete_sequence=["#00A08B"],
             title='Constructor Standings for 2021',
             labels={'percentage': 'Performance Percentage'},
             height=600, width=800)

# Update x-axis title and fonts
fig.update_xaxes(title_text='Distribution Percentage',title_font=dict(size=18))

# Update y-axis title
fig.update_yaxes(title_text='Constructor Name',title_font=dict(size=18))

# Update marker settings
fig.update_traces(marker=dict(line=dict(width=1, color='black')))

# Update title font size
fig.update_layout(title_font=dict(size=24))


# Show the plot
fig.show()

In [25]:
constructor_standings_2021_to_2022 = merged_constructor_races_df[
    (merged_constructor_races_df['year'] == 2022)
]

# Group the data by constructor_name and calculate the mean position for each constructor
constructor_standings_mean = constructor_standings_2021_to_2022.groupby('constructor_name')['position'].mean()

# Sort the data in ascending order based on mean position
constructor_standings_mean = constructor_standings_mean.sort_values(ascending=False)  # Reverse the order

# Calculate the percentage relative to the best possible mean position (100% for the best)
best_mean_position = constructor_standings_mean.min()
constructor_standings_percentage = 100 * (best_mean_position / constructor_standings_mean)

# Create a DataFrame with constructor names and mean positions in percentage
data = pd.DataFrame({'constructor_name': constructor_standings_mean.index, 'percentage': constructor_standings_percentage})

# Create a bar chart using Plotly Express
fig = px.bar(data, x='percentage', y='constructor_name', orientation='h',
             color_discrete_sequence=["#00A08B"],
             title='Constructor Standings for 2022',
             labels={'percentage': 'Performance Percentage'},
             height=600, width=800)

# Update x-axis title and fonts
fig.update_xaxes(title_text='Distribution Percentage',title_font=dict(size=18))

# Update y-axis title
fig.update_yaxes(title_text='Constructor Name',title_font=dict(size=18))

# Update marker settings
fig.update_traces(marker=dict(line=dict(width=1, color='black')))

# Update title font size
fig.update_layout(title_font=dict(size=24))


# Show the plot
fig.show()

In [26]:
import pandas as pd
import plotly.express as px

# Assuming your DataFrame is named merged_constructor_races_df

# Filter the DataFrame to include data for the years 2018 to 2022
constructor_standings_2018_to_2022 = merged_constructor_races_df[
    (merged_constructor_races_df['year'] >= 2018) & (merged_constructor_races_df['year'] <= 2022)
]

# Group the data by constructor_name and calculate the mean position for each constructor
constructor_standings_mean = constructor_standings_2018_to_2022.groupby('constructor_name')['position'].mean()

# Sort the data in ascending order based on mean position
constructor_standings_mean = constructor_standings_mean.sort_values(ascending=False)  # Reverse the order

# Calculate the percentage relative to the best possible mean position (100% for the best)
best_mean_position = constructor_standings_mean.min()
constructor_standings_percentage = 100 * (best_mean_position / constructor_standings_mean)

# Create a DataFrame with constructor names and mean positions in percentage
data = pd.DataFrame({'constructor_name': constructor_standings_mean.index, 'percentage': constructor_standings_percentage})

# Create a bar chart using Plotly Express
fig = px.bar(data, x='percentage', y='constructor_name', orientation='h',
             color_discrete_sequence=["#00A08B"],
             title='Constructor Standings for 2018-2022',
             labels={'percentage': 'Performance Percentage'},
             height=600, width=800)

# Update x-axis title and fonts
fig.update_xaxes(title_text='Distribution Percentage', title_font=dict(size=18))

# Update y-axis title
fig.update_yaxes(title_text='Constructor Name', title_font=dict(size=18))

# Update marker settings
fig.update_traces(marker=dict(line=dict(width=1, color='black')))

# Update title font size
fig.update_layout(title_font=dict(size=24))

# Show the plot
fig.show()


In [27]:
merged_constructor_races_df.tail()

raceId  constructorId  points  position  wins constructor_name  year   
12936     679            183     2.0         5     0   Brabham-Climax  1967  \
12937     679            170     6.0         2     0    Cooper-Climax  1967   
12938     679            176     0.0         7     0        Lotus-BRM  1967   
12939     679            184     0.0         6     0       LDS-Climax  1967   
12940     679            189     0.0         8     0     Eagle-Climax  1967   

       circuitId              circuit_name  
12936         30  South African Grand Prix  
12937         30  South African Grand Prix  
12938         30  South African Grand Prix  
12939         30  South African Grand Prix  
12940         30  South African Grand Prix

In [28]:
merged_constructor_races_df.year.unique()

array([2008, 2007, 2006, 2005, 2004, 2003, 2002, 2001, 2000, 1999, 1998,
       1997, 1996, 1995, 1994, 1993, 1992, 1990, 2009, 1991, 1989, 1988,
       1987, 1986, 1985, 1984, 1983, 1982, 1981, 1980, 1979, 1978, 1977,
       1976, 1975, 1974, 1973, 1972, 1971, 1968, 2010, 2011, 2012, 2013,
       2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 1967,
       1966, 1965, 1963, 1964, 1962, 1961, 1960, 1959, 1958, 1970, 1969])

In [29]:
filtered_df = merged_constructor_races_df[merged_constructor_races_df['year'] == 2023]

# Now, filtered_df contains only the rows where the 'year' column is equal to 2023

filtered_df.head(40)

raceId  constructorId  points  position  wins constructor_name  year   
11365    1098              1     0.0         7     0          McLaren  2023  \
11366    1098              3     0.0        10     0         Williams  2023   
11367    1098              6     0.0         5     0          Ferrari  2023   
11368    1098              9     0.0         9     0         Red Bull  2023   
11369    1098             51     0.0         1     0       Alfa Romeo  2023   
11370    1098            117     0.0         4     0     Aston Martin  2023   
11371    1098            131     0.0         8     0         Mercedes  2023   
11372    1098            210     0.0         6     0     Haas F1 Team  2023   
11373    1098            213     0.0         2     0       AlphaTauri  2023   
11374    1098            214     0.0         3     0   Alpine F1 Team  2023   

       circuitId        circuit_name  
11365          3  Bahrain Grand Prix  
11366          3  Bahrain Grand Prix  
11367          3  Bahrain Grand Prix  
11368          3  Bahrain Grand Prix  
11369          3  Bahrain Grand Prix  
11370          3  Bahrain Grand Prix  
11371          3  Bahrain Grand Prix  
11372          3  Bahrain Grand Prix  
11373          3  Bahrain Grand Prix  
11374          3  Bahrain Grand Prix

Unfortunately our dataframe does not have updated data regarding 2023. 

According to the [Fia](https://www.formula1.com/en/results.html/2023/team.html) website, these are the Constructor standings results so far (11.09.2023) :


| POS | TEAM                              | PTS |
|-----|-----------------------------------|-----|
| 1   | RED BULL RACING HONDA RBPT       | 583 |
| 2   | MERCEDES                          | 273 |
| 3   | FERRARI                           | 228 |
| 4   | ASTON MARTIN ARAMCO MERCEDES     | 217 |
| 5   | MCLAREN MERCEDES                  | 115 |
| 6   | ALPINE RENAULT                    | 73  |
| 7   | WILLIAMS MERCEDES                 | 21  |
| 8   | HAAS FERRARI                      | 11  |
| 9   | ALFA ROMEO FERRARI                | 10  |
| 10  | ALPHATAURI HONDA RBPT             | 3   |


Let's create a plot to have a better overview:

In [30]:
# Team names and points
teams = [
    'RED BULL RACING HONDA RBPT', 'MERCEDES', 'FERRARI', 'ASTON MARTIN ARAMCO MERCEDES',
    'MCLAREN MERCEDES', 'ALPINE RENAULT', 'WILLIAMS MERCEDES', 'HAAS FERRARI',
    'ALFA ROMEO FERRARI', 'ALPHATAURI HONDA RBPT'
]
points = [583, 273, 228, 217, 115, 73, 21, 11, 10, 3]

# Create a DataFrame for the data
data = {'Team': teams, 'Points': points}

# Create a bar plot using Plotly Express
fig = px.bar(data, x='Team', y='Points', color_discrete_sequence=["#00A08B"],
             labels={'Points': 'Total Points'},
             text=points)

# Set the title with a shorter length
fig.update_layout(
    title='Formula 1 Team Points Breakdown 2023',  # Catchy title
    title_font=dict(size=24),  # Set title font size to 24
    xaxis=dict(title='Team', title_font=dict(size=18)),
    yaxis=dict(title='Total Points', title_font=dict(size=18)),
    height=600,  # Set the height to 600
    width=800    # Set the width to 800
)

# Show the plot
fig.show()


And these are according to the [Fia](https://chat.openai.com/?model=text-davinci-002-render-sha) website, these are the Drivers standings results so far (11.09.2023) :

| POS | DRIVER              | NATIONALITY | CAR                                 | PTS |
|-----|---------------------|-------------|-------------------------------------|-----|
| 1   | Max Verstappen      | NED         | RED BULL RACING HONDA RBPT         | 364 |
| 2   | Sergio Perez        | MEX         | RED BULL RACING HONDA RBPT         | 219 |
| 3   | Fernando Alonso     | ESP         | ASTON MARTIN ARAMCO MERCEDES       | 170 |
| 4   | Lewis Hamilton      | GBR         | MERCEDES                            | 164 |
| 5   | Carlos Sainz        | ESP         | FERRARI                             | 117 |
| 6   | Charles Leclerc     | MON         | FERRARI                             | 111 |
| 7   | George Russell      | GBR         | MERCEDES                            | 109 |
| 8   | Lando Norris        | GBR         | MCLAREN MERCEDES                    | 79  |
| 9   | Lance Stroll        | CAN         | ASTON MARTIN ARAMCO MERCEDES       | 47  |
| 10  | Pierre Gasly        | FRA         | ALPINE RENAULT                      | 37  |
| 11  | Esteban Ocon        | FRA         | ALPINE RENAULT                      | 36  |
| 12  | Oscar Piastri       | AUS         | MCLAREN MERCEDES                    | 36  |
| 13  | Alexander Albon     | THA         | WILLIAMS MERCEDES                   | 21  |
| 14  | Nico Hulkenberg     | GER         | HAAS FERRARI                        | 9   |
| 15  | Valtteri Bottas     | FIN         | ALFA ROMEO FERRARI                  | 6   |
| 16  | Zhou Guanyu         | CHN         | ALFA ROMEO FERRARI                  | 4   |
| 17  | Yuki Tsunoda        | JPN         | ALPHATAURI HONDA RBPT               | 3   |
| 18  | Kevin Magnussen     | DEN         | HAAS FERRARI                        | 2   |
| 19  | Logan Sargeant      | USA         | WILLIAMS MERCEDES                   | 0   |
| 20  | Liam Lawson         | NZL         | ALPHATAURI HONDA RBPT               | 0   |
| 21  | Nyck De Vries       | NED         | ALPHATAURI HONDA RBPT               | 0   |
| 22  | Daniel Ricciardo    | AUS         | ALPHATAURI HONDA RBPT               | 0   |


Fernando Alonso has 170 points and Lance Stroll has 47.

That means:

For Fernando Alonso:
Percentage Alonso = (170 / 217) * 100 ≈ 78.34%

For Lance Stroll:
Percentage Stroll = (47 / 217) * 100 ≈ 21.66%

Fernando Alonso contributes with 78% of the total points Aston Martin has.
It's a huge difference comparing with his team mate, Lance Stroll.

Let's create a plot from these numbers:


In [31]:
# Driver names and points
drivers = ['Fernando Alonso', 'Lance Stroll']
points = [170, 47]

# Calculate the percentages
total_points = sum(points)
percentages = [(p / total_points) * 100 for p in points]

# Create a DataFrame for the data
data = {'Driver': drivers, 'Percentage of Total Points (%)': percentages}

# Create a bar plot using Plotly Express
fig = px.bar(data, x='Driver', y='Percentage of Total Points (%)', color_discrete_sequence=["#00A08B"],
             labels={'Percentage of Total Points (%)': 'Percentage of Total Points (%)'},
             text=percentages)

# Format the text labels as percentages with no decimal places
fig.update_traces(texttemplate='%{text:.0f}%')

# Set the y-axis range to 0-100%
fig.update_yaxes(range=[0, 100])

# Increase the font size for the title and axis labels
fig.update_layout(
    title="Aston Martin's Dynamic Duo: Points Breakdown 2023",
    title_font=dict(size=24),  # Set title font size to 24
    xaxis=dict(title_font=dict(size=18)),
    yaxis=dict(title_font=dict(size=18)),
    height=600,  # Set the height to 600
    width=800    # Set the width to 800
)

# Show the plot
fig.show()


Comparing drivers Fernando Alonso, Lance Stroll in season 2023.

According to the [Racing Statistics](https://www.racing-statistics.com/en/f1-drivers/compare/fernando-alonso/lance-stroll/seasons/2023) website, these are the numbers so far (11.09.2023) :

|                            | Fernando Alonso Aston Martin | Lance Stroll Aston Martin |
|----------------------------|-----------------------------|---------------------------|
| **Events**                 | 14                          | 14                        |
| **Championship result**    | 3rd                         | 9th                       |
| **Best race result**       | 2nd                         | 4th                       |
| **Wins**                   | 0                           | 0                         |
| **Pole positions**         | 0                           | 0                         |
| **Fastest laps**           | 1                           | 0                         |
| **Points**                 | 170                         | 47                        |
| **Laps**                   | 890                         | 839                       |
| **DNFs**                   | 0                           | 2                         |
| **Teammate**               | Lance Stroll                | Fernando Alonso           |
| **Average finish position**| 4.57                        | 9.42                      |
| **Average grid position**  | 5.64                        | 11.00                     |


In [32]:
# Driver names
drivers = ['Fernando Alonso', 'Lance Stroll']

# Average finish position and average grid position data
average_finish_position = [4.57, 9.42]
average_grid_position = [5.64, 11.00]

# Create a DataFrame for the data
data = {
    'Driver': drivers,
    'Average Finish Position': average_finish_position,
    'Average Grid Position': average_grid_position
}

# Create a grouped bar plot without stacking using Plotly Express
fig = px.bar(data, x='Driver', y=['Average Finish Position', 'Average Grid Position'],
             color_discrete_map={'Average Finish Position': 'rgb(102,194,165)', 'Average Grid Position': 'darkgreen'},
             labels={'value': 'Position'},
             title='Average Finish and Grid Positions for Fernando Alonso and Lance Stroll (2023)',
             barmode='group')

# Set the y-axis range
fig.update_yaxes(range=[0, 15], title_font=dict(size=18))  # Set y-axis title font size

# Set the x-axis title font size
fig.update_xaxes(title_font=dict(size=18))

# Set the title font size
fig.update_layout(title_text='Finish vs. Grid Positions for Fernando Alonso and Lance Stroll (2023)',
                  title_font=dict(size=24))  # Set title font size

# Show the plot
fig.show()


In [33]:
# Import necessary libraries
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Driver names
drivers = ['Fernando Alonso', 'Lance Stroll']

# Average finish position and average grid position data
average_finish_position = [4.57, 9.42]
average_grid_position = [5.64, 11.00]

# Create subplots with two rows and one column
fig = make_subplots(rows=2, cols=1, subplot_titles=['Average Grid Positions (2023)', 'Average Finish Positions (2023)'])

# Add the first bar plot for average grid positions
fig.add_trace(go.Bar(x=drivers, y=average_grid_position, marker_color='darkgreen', name='Average Grid Position'), row=1, col=1)

# Add the second bar plot for average finish positions
fig.add_trace(go.Bar(x=drivers, y=average_finish_position, marker_color='rgb(102,194,165)', name='Average Finish Position'), row=2, col=1)

# Update y-axis titles and ranges
fig.update_yaxes(title_text='Grid Position', range=[0, 15], row=1, col=1)
fig.update_yaxes(title_text='Finish Position', range=[0, 15], row=2, col=1)

# Update x-axis title
fig.update_xaxes(title_text='Driver')

# Update the overall layout
fig.update_layout(
    title_text='Average Grid and Finish Positions for Fernando Alonso and Lance Stroll (2023)',
    title_font=dict(size=24),
    barmode='group'
)

# Show the plot
fig.show()


In [34]:
# Import necessary libraries
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Driver names
drivers = ['Fernando Alonso', 'Lance Stroll']

# Average finish position and average grid position data
average_finish_position = [4.57, 9.42]
average_grid_position = [5.64, 11.00]

# Create subplots with two rows and one column
fig = make_subplots(rows=2, cols=1, subplot_titles=['Average Grid Positions (2023)', 'Average Finish Positions (2023)'])

# Add the first bar plot for average grid positions with text labels
fig.add_trace(go.Bar(x=drivers, y=average_grid_position, marker_color='darkgreen', name='Average Grid Position',
                     text=average_grid_position, textposition='inside'), row=1, col=1)

# Add the second bar plot for average finish positions with text labels
fig.add_trace(go.Bar(x=drivers, y=average_finish_position, marker_color='rgb(102,194,165)', name='Average Finish Position',
                     text=average_finish_position, textposition='inside'), row=2, col=1)

# Update y-axis titles and ranges
fig.update_yaxes(title_text='Grid Position', range=[0, 15], row=1, col=1)
fig.update_yaxes(title_text='Finish Position', range=[0, 15], row=2, col=1)

# Update x-axis title
fig.update_xaxes(title_text='Driver')

# Update the overall layout
fig.update_layout(
    title_text='Average Grid and Finish Positions for Fernando Alonso and Lance Stroll (2023)',
    title_font=dict(size=24),
    barmode='group'
)

# Show the plot
fig.show()


In [37]:
# Import necessary libraries
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Driver names
drivers = ['Fernando Alonso', 'Lance Stroll']

# Average finish position and average grid position data
average_finish_position = [4.57, 9.42]
average_grid_position = [5.64, 11.00]

# Create subplots with two rows and one column
fig = make_subplots(rows=2, cols=1, subplot_titles=['Average Grid Positions', 'Average Finish Positions'])

# Format the values to one decimal place
formatted_average_grid_position = [f'{val:.0f}' for val in average_grid_position]
formatted_average_finish_position = [f'{val:.0f}' for val in average_finish_position]

# Add the first bar plot for average grid positions with formatted text labels
fig.add_trace(go.Bar(x=drivers, y=average_grid_position, marker_color='darkgreen', name='Average Grid Position',
                     text=formatted_average_grid_position, textposition='inside'), row=1, col=1)

# Add the second bar plot for average finish positions with formatted text labels
fig.add_trace(go.Bar(x=drivers, y=average_finish_position, marker_color='rgb(102,194,165)', name='Average Finish Position',
                     text=formatted_average_finish_position, textposition='inside'), row=2, col=1)

# Update y-axis titles and ranges
fig.update_yaxes(title_text='Grid Position', range=[0, 15], row=1, col=1)
fig.update_yaxes(title_text='Finish Position', range=[0, 15], row=2, col=1)

# Update x-axis title
fig.update_xaxes(title_text='Driver')

# Update the overall layout
fig.update_layout(
    title_text='Average Grid and Finish Positions for Fernando Alonso and Lance Stroll (2023)',
    title_font=dict(size=24),
    barmode='group'
)

# Show the plot
fig.show()
